Add support for MAX17616

cjones-adi opened the pull request:
## Pull Request Description

The MAX17616/MAX17616A offers highly versatile and programmable protection boundaries for systems against input voltage faults and output overcurrent faults. Input-voltage faults (with positive polarity) are protected up to +80V (without Reverse Current Protection)/+75V (with Reverse Current Protection), by an internal nFET featuring low ON-resistance (20mΩ typ). The devices feature a programmable undervoltage-lockout (UVLO) thresholds by using external voltage- dividers. The MAX17616 features a programmable overvoltage-lockout (OVLO) while MAX17616A offers a programmable output voltage clamp function through the OVFB pin that features an output voltage limiting regulation during input transient surge events. Input undervoltage and overvoltage protection (MAX17616)/output voltage clamp function (MAX17616A) can be programmed across the entire 3V to 80V operating range.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have complied with the [Submission Checklist](http://analogdevicesinc.github.io/no-OS/contributing.html#submission-checklist)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2734) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2734) before we can accept your contribution.<br/><hr/>**Carlos Jones** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=2734) it.</sub>

cjones-adi commented:
V2:

Rebase to latest main
Moved max17616.rst to respective doc/sphinx/source directory
Fix astyle errors on identified files

kister-jimenez commented:
/AzurePipelines run

jemfgeronimo commented:
/AzurePipelines run

cjones-adi commented:
V4:

Remove diff headers included in max17616.h

cjones-adi commented:
/AzurePipelines run

jemfgeronimo commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1191 @@
+		telemetry->valid_mask |= NO_OS_BIT(3);
+
+	/* Read Temperature using DIRECT format */
+	ret = max17616_read_value(dev, MAX17616_TEMP, &telemetry->temp1);
```
Comments are redundant; the function calls already describe what’s happening.

cjones-adi replied:
Thanks. Fixed in latest commit.

jemfgeronimo commented on drivers/power/max17616/max17616.c:
```diff
@@ -563,8 +563,8 @@ int max17616_get_operation_state(struct max17616_dev *dev, bool *enabled)
 
-	int ret = max17616_read_byte(dev,MAX17616_CMD(MAX17616_OPERATION),
-				   &operation);
+	int ret = max17616_read_byte(dev, MAX17616_CMD(MAX17616_OPERATION),
```
Include `max17616.c` and `max17616.h` changes to the previous commit.

cjones-adi replied:
Thanks. Fixed in latest commit.

jemfgeronimo commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,771 @@
+	int value;
+	int ret;
+
+	/* Parse the input value */
```
Unnecessary comment.

cjones-adi replied:
Thanks. Fixed in latest commit.

cjones-adi commented:
V5:

- Improve error handling on max17616.c.
- Move variable declaration to top of function.
- Use macros in switch cases.
- Remove explicit value assignments in enum declarations.
- Change to new example infrastructure.
- Remove redundant comments on self-explanatory function names.
- Remove magic numbers by adding descriptive macros.
- Fix file content in commit.

cjones-adi commented on drivers/power/max17616/max17616.h:
```diff
@@ -0,0 +1,439 @@
+	MAX17616_STATUS_BIT_IOUT_OC	= 4,	// Output Over Current Fault
+	MAX17616_STATUS_BIT_VOUT_OV	= 5,	// An output overvoltage fault occurred
+	MAX17616_STATUS_BIT_OFF		= 6,	// Max17616 is not enabled
+	MAX17616_STATUS_BIT_BUSY	= 7	// Not supported
```
Thanks. Fixed in latest commit.

cjones-adi commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1191 @@
+		return ret;
+
+	/* Step 3: Verify PMBus revision */
+	ret = max17616_verify_pmbus_revision(dev);
```
Thanks. Fixed in latest commit.

cjones-adi commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,771 @@
+
+		return snprintf(buf, len, "%s", fault_info);
+	}
+	case 4: { /* capability */
```
Thanks. Fixed in latest commit.

cjones-adi commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,771 @@
+		return max17616_set_operation_state(iio_max17616->max17616_dev, enable);
+	}
+
+	case 1: /* clear_faults */
```
Thanks. Fixed in latest commit.

cjones-adi commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,780 @@
+	{
+		.name = "scale",
+		.show = max17616_iio_read_attr,
+	},
```
Yes, it returns the same data. For this case, with specs to use DIRECT method of transforming data, returned values are in V, I, or degC units already (similar cases are devices LTC4162L and LT8491 where the read values are in engineering units already and the scale factor used is 1). I'll use the `.priv` field and scale factor 1 for IIO compliance.

cjones-adi replied:
Thanks. Addressed in latest commit.

jemfgeronimo commented:
LGTM

jemfgeronimo commented:
/AzurePipelines run

kister-jimenez commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,818 @@
+	MAX17616_IIO_FAULT_SUMMARY_ATTR,
+	MAX17616_IIO_CAPABILITY_ATTR,
+};
+
```
NIT. extra newline.

cjones-adi commented:
V9:

- Replace returned data of DIRECT conversions from int to float to preserve precision.

kister-jimenez commented:
/AzurePipelines run

amiclaus commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,818 @@
+	if (!iio_desc)
+		return -ENODEV;
+
+	if (iio_desc->iio_dev)
```
`iio_desc->max17616_dev` ?

cjones-adi replied:
Corrected the struct member to validate

amiclaus commented on drivers/power/max17616/max17616.h:
```diff
@@ -0,0 +1,485 @@
+	MAX17616_VIN,		/* Input voltage in volts */
+	MAX17616_VOUT,		/* Output voltage in volt */
+	MAX17616_IOUT,		/* Output current in amps */
+	MAX17616_TEMP,		/* Temperature in degreesCelsius */
```
nit: `degrees Celsius`

amiclaus commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1178 @@
+		return -EINVAL;
+
+	return max17616_write_byte(dev, MAX17616_CMD(MAX17616_SET_ISTART_RATIO),
+				   (uint8_t)istart_ratio);
```
any `istart_ratio` value is valid?

cjones-adi replied:
Provided the default values as per datasheet in the initialization

cjones-adi commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1178 @@
+{
+	uint16_t raw_value;
+	int ret;
+	float vin, iout;
```
Replaced all float instances with uint32_t and adjusted calculations

cjones-adi commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1178 @@
+		return -EINVAL;
+
+	/* Combine voltage selection (bits 4:2) and PGOOD threshold (bits 1:0) */
+	reg_value = ((uint8_t)voltage << 2) | (uint8_t)threshold;
```
Used no-OS field preparation macros

amiclaus commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1229 @@
+ * Since R = -1 for all MAX17616 coefficients:
+ * X_milli = ((Y x 10 - b) x 1000) / m
+ */
+static float max17616_direct_to_milliunit(uint16_t raw_value,
```
why float if it returns an int32_t ?

amiclaus commented on tests/drivers/power/max17616/test/test_max17616.c:
```diff
@@ -0,0 +1,1605 @@
+	no_os_i2c_read_IgnoreAndReturn(0);
+	no_os_get_unaligned_le16_IgnoreAndReturn(0x2000);  // IOUT value
+
+	result = max17616_read_value(&device, MAX17616_POWER, &value);
```
ditto.

cjones-adi replied:
Updated test variable data type in this test case.

amiclaus commented on tests/drivers/power/max17616/test/test_max17616.c:
```diff
@@ -0,0 +1,1605 @@
+	float value = 0.0f;
+
+	// Test with NULL device
+	int result = max17616_read_value(NULL, MAX17616_VIN, &value);
```
same. please check all occurences where u pass float to a function that expects int32_t

amiclaus commented on tests/drivers/power/max17616/test/support/test_max17616_support.c:
```diff
@@ -0,0 +1,374 @@
+	}
+
+	if (telemetry->valid_mask & 0x08) { /* IOUT */
+		if (telemetry->iout_ma < 0 || telemetry->iout_ma > 1000) { /* 0-100A */
```
ditto.

cjones-adi replied:
Updated stale test cases from float to int32_t conversion

cjones-adi commented on tests/drivers/power/max17616/test/test_max17616.c:
```diff
@@ -0,0 +1,1605 @@
+	TEST_ASSERT_EQUAL_INT(-EINVAL, result);
+
+	// Test with invalid value type
+	result = max17616_read_value(&device, 999, &value);
```
Updated test variable data type in this test case.

cjones-adi commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1229 @@
+
+	case MAX17616_POWER:
+		/* Calculate power from voltage and current: P = V × I */
+		ret = max17616_read_value(dev, MAX17616_VOUT, &vout_mv);
```
Aligned data type. Readjusted power calculation to handle overflow.

cjones-adi commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,818 @@
+
+	switch (channel->address) {
+
+	case MAX17616_IIO_CLMODE_CHAN:
```
Added write attribute switch cases for NOMINAL_VOLTAGE_CHAN and PGOOD_THRESHOLD_CHAN

cjones-adi commented on projects/max17616/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,393 @@
+	struct max17616_telemetry telemetry;
+	int ret;
+
+	ret = no_os_uart_init(&uart_desc, &uart_ip);
```
Added clean-up of uart_desc on error path

cjones-adi commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,818 @@
+	struct max17616_iio_desc *iio_max17616 = (struct max17616_iio_desc *)device;
+	struct max17616_telemetry telemetry;
+	struct max17616_status status;
+	uint16_t raw_value;
```
Removed unused variable

cjones-adi commented:
V13

- Rebased to latest main
- Addressed PR comments (Update overlooked float function return and unit test cases from float to int32_t conversion, add NULL check on max17616_read_status_byte(), align local variable type qualifiers for power calculation, add missing iio write attribute for nominal voltage and pgood threshold channels, etc.)

amiclaus commented on drivers/power/max17616/max17616.c:
```diff
@@ -0,0 +1,1233 @@
+int max17616_remove(struct max17616_dev *dev)
+{
+	int ret;
+
```
missing NULL check for dev.

amiclaus commented on tests/drivers/power/max17616/test/test_iio_max17616.c:
```diff
@@ -0,0 +1,2195 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+ * EVENT SHALL ANALOG DEVICES, INC. BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR SERVICES; LOSS OF USE, DATA,
```
part of the header is missing.

amiclaus commented on drivers/power/max17616/iio_max17616.c:
```diff
@@ -0,0 +1,855 @@
+};
+
+/* IIO device structure - declared early for use in init function */
+static struct iio_device max17616_iio_device = {
```
do we want this to be static?

amiclaus commented on projects/max17616/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,393 @@
+		/* Read telemetry */
+		ret = max17616_read_telemetry_all(max17616_dev, &telemetry);
+		if (ret)
+			pr_err("Failed to read telemetry: %d\n\r", ret);
```
we should somehow go to exit: for proper cleanup

cjones-adi replied:
looking at max17616_read_telemetry_all, it doesn't really return error worthy of discontinuing the loop. modified the max17616_read_telemetry_all to catch read operation errors so it would warrant an exit and clean up here. great observations. thanks.

amiclaus reviewed (commented):
some comments on my side.

stefpopa commented on projects/max17616/builds.json:
```diff
@@ -0,0 +1,10 @@
+			"flags" : "EXAMPLE=iio_example TARGET=max32665"
+		}
+	}
+}
```
is this a typo? I believe it's max32665 not max32666?

stefpopa reviewed (commented):
A few minor remarks, but overall looks good to me.

cjones-adi commented on doc/sphinx/source/projects/power/max17616.rst:
```diff
@@ -0,0 +1 @@
+.. include:: ../../../../../projects/max17616/README.rst
```
Thanks. Newline at EOF here is not standardized though. I'll add anyway.
